### PR TITLE
Gpu vectors

### PR DESCRIPTION
**Pull request by IvanPleshkov**
This PR introduced GPU storing and scoring of vectors. It supports dense vectors with `f32, f16, u8` type, SQ/BQ/PQ + multivectors for all types and quantization.

I marker in the comments all pieces of changes that are not under GPU-feature.

### CODE REVIEW

File: lib/segment/src/index/hnsw_index/gpu/gpu_vector_storage/tests.rs [hunk 1]
```diff
@@ -0,0 +1,678 @@
+use std::path::Path;
```
Comment (IvanPleshkov) [inline lib/segment/src/index/hnsw_index/gpu/gpu_vector_storage/tests.rs:1]: A file with all unit tests: f32, f16, u8 dense/multivectors with SQ, BQ, PQ.

File: lib/quantization/src/encoded_vectors_pq.rs [hunk 2]
```diff
@@ -33,10 +33,10 @@ pub struct EncodedQueryPQ {
 }
 
 #[derive(Serialize, Deserialize)]
-struct Metadata {
-    centroids: Vec<Vec<f32>>,
-    vector_division: Vec<Range<usize>>,
-    vector_parameters: VectorParameters,
+pub struct Metadata {
```
Comment (IvanPleshkov) [inline lib/quantization/src/encoded_vectors_pq.rs:36]: Made it public to read this parameters and copy to the gpu

File: lib/quantization/src/encoded_vectors_pq.rs [hunk 3]
```diff
@@ -423,6 +423,19 @@ impl<TStorage: EncodedStorage> EncodedVectorsPQ<TStorage> {
             })
             .sum()
     }
+
+    pub fn get_quantized_vector(&self, i: u32) -> &[u8] {
```
Comment (IvanPleshkov) [inline lib/quantization/src/encoded_vectors_pq.rs:427]: Get quantized vector to copy to the GPU + parameter getters

File: lib/quantization/src/encoded_vectors_binary.rs [hunk 4]
```diff
@@ -254,6 +254,19 @@ impl<TBitsStoreType: BitsStoreType, TStorage: EncodedStorage>
             (DistanceType::L1 | DistanceType::L2, false) => xor_product - zeros_count,
         }
     }
+
+    pub fn get_quantized_vector(&self, i: u32) -> &[u8] {
```
Comment (IvanPleshkov) [inline lib/quantization/src/encoded_vectors_binary.rs:258]: Get quantized vector to copy to the GPU + parameter getters

File: lib/segment/src/common/operation_error.rs [hunk 5]
```diff
@@ -192,6 +192,13 @@ impl From<TryReserveError> for OperationError {
     }
 }
 
+#[cfg(feature = "gpu")]
+impl From<gpu::GpuError> for OperationError {
```
Comment (IvanPleshkov) [inline lib/segment/src/common/operation_error.rs:196]: Convertion from gpu error

File: lib/segment/src/vector_storage/quantized/quantized_multivector_storage.rs [hunk 6]
```diff
@@ -201,6 +215,18 @@ where
         }
         sum
     }
+
+    pub fn inner_storage(&self) -> &QuantizedStorage {
```
Comment (IvanPleshkov) [inline lib/segment/src/vector_storage/quantized/quantized_multivector_storage.rs:219]: Added getters into quantized multivectors. Function to get original dense quantized vectors and some paramenets

File: lib/segment/src/index/hnsw_index/gpu/gpu_vector_storage/tests.rs (hunk 1 above)
Comment (timvisee) [reply lib/segment/src/index/hnsw_index/gpu/gpu_vector_storage/tests.rs:1]: Maybe we can add this as a comment to the module.

```suggestion
//! Unit tests for GPU vector storage, covering f32, f16, u8 dense/multivectors with SQ, BQ, PQ.

use std::path::Path;
```

File: lib/quantization/src/encoded_vectors_binary.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (IvanPleshkov) [reply lib/quantization/src/encoded_vectors_binary.rs:258]: To do it, I have to add `gpu` feature into quantization crate. I don't want to do it just for getters

[the remark this answers is not in the document]
Comment (timvisee) [reply lib/quantization/src/encoded_vectors_binary.rs:258]: Makes sense :+1:

File: lib/segment/src/index/hnsw_index/gpu/mod.rs [hunk 7]
```diff
@@ -0,0 +1,4 @@
+pub mod gpu_vector_storage;
+pub mod shader_builder;
+
+static GPU_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(60);
```
[the remark this answers is not in the document]
Comment (IvanPleshkov) [reply lib/segment/src/index/hnsw_index/gpu/mod.rs:8]: added a comment

**Review by timvisee — APPROVED** [review]

**Review by generall — APPROVED** [review]

**Pull request merged**